File: applications/carrot2-dcs/examples/csharp/src/Examples.cs

using System;
using System.Reflection;
using System.IO;
using System.Xml.Serialization;
using System.Xml;
using System.Xml.XPath;
using System.Collections;

/// 
/// <summary>
/// Contains several examples of clustering using the Document Clustering Server.
/// </summary>
///
namespace Org.Carrot2.Examples
{
    public sealed class Examples
    {
        public static void Main()
        {
            MultipartFileUpload service = new MultipartFileUpload(new Uri("http://localhost:8080/dcs/rest"));

            string examplePath = "..\\..\\..\\shared\\data-mining.xml";
            if (!File.Exists(examplePath))
            {
                Console.WriteLine("Input path does not exist: " + examplePath);
                return;
            }

            // Cluster directly from file (no buffering).
            Console.WriteLine("## Clustering documents from a file...");
            ClusterFromFile(service, "..\\..\\..\\shared\\data-mining.xml", "data mining");

            // Cluster from an XML in memory.
            string xml = File.ReadAllText(examplePath, System.Text.Encoding.UTF8);
            Console.WriteLine("## Clustering documents from an XML string...");
            ClusterFromStream(service, 
                new MemoryStream(System.Text.Encoding.UTF8.GetBytes(xml)), "data mining");

            // Cluster form an external document source (on the DCS).
            Console.WriteLine("## Clustering search results from a search engine...");
            ClusterFromSearchEngine(service, "etools", "data mining");
        }

        /// <summary>
        /// An example of clustering data from an arbitrary byte stream holding input XML for
        /// the DCS (can be an in-memory stream if clustering from a string).
        /// </summary>
        private static void ClusterFromStream(MultipartFileUpload service, Stream xmlStream, string queryHint)
        {
            // The output format is XML.
            service.AddFormValue("dcs.output.format", "XML");

            // We don't need documents in the output, only clusters.
            service.AddFormValue("dcs.clusters.only", "true");

            // Pass query hint.
            service.AddFormValue("query", queryHint);

            // The algorithm to use for clustering. Omit to select the default. An example of
            // using Lingo with custom parameters follows.

            service.AddFormValue("dcs.algorithm", "lingo");
            service.AddFormValue("LingoClusteringAlgorithm.desiredClusterCountBase", "10");
            service.AddFormValue("LingoClusteringAlgorithm.factorizationQuality", "LOW");
            service.AddFormValue("LingoClusteringAlgorithm.factorizationFactory",
                "org.carrot2.matrix.factorization.PartialSingularValueDecompositionFactory");

            // Add the XML stream here.
            service.AddFormStream("dcs.c2stream", "anything.xml", xmlStream);

            // Perform the actual query.
            byte[] response = service.Post();

            // Parse the output and dump group headers.
            MemoryStream input = new MemoryStream(response);

            XmlDocument document = new XmlDocument();
            document.PreserveWhitespace = true;
            document.Load(input);

            PrintResults(document);
        }

        /// <summary>
        /// An example of clustering data stored in a local file and passed
        /// as part of the HTTP request. A query hint is provided for the
        /// clustering algorithm (to avoid trivial clusters).
        /// </summary>
        private static void ClusterFromFile(MultipartFileUpload service, string filePath, string queryHint)
        {
            using (FileStream fs = File.Open(filePath, FileMode.Open))
            {
                ClusterFromStream(service, fs, queryHint);
            }
        }

        /// <summary>
        /// Cluster data retrieved from a search engine or some other source registered in the
        /// DCS as a document source.
        /// </summary>
        private static void ClusterFromSearchEngine(MultipartFileUpload service, string sourceId, string query)
        {
            // The output format is XML.
            service.AddFormValue("dcs.output.format", "XML");

            // This time we will be interested in both clusters and documents.
            service.AddFormValue("dcs.clusters.only", "false");

            // Add query.
            service.AddFormValue("query", query);

            // Add the number of results.
            service.AddFormValue("results", "20");

            // Specify the source.
            service.AddFormValue("dcs.source", sourceId);

            // Perform the actual query.
            byte[] response = service.Post();

            // Parse the output and dump group headers.
            MemoryStream input = new MemoryStream(response);

            XmlDocument document = new XmlDocument();
            document.PreserveWhitespace = true;
            document.Load(input);

            PrintResults(document);
        }

        /// <summary>
        /// Dump the result (group labels).
        /// </summary>
        private static void PrintResults(XmlDocument document)
        {
            foreach (XmlNode group in
                document.SelectNodes("/searchresult/group"))
            {
                PrintGroup(group, 1);
            }
            Console.WriteLine();
        }

        /// <summary>
        /// Dump a single cluster and its subclusters.
        /// </summary>
        private static void PrintGroup(XmlNode group, int level)
        {
            for (int i = 0; i < level; i++) Console.Write("  ");
            Console.Write(group.SelectSingleNode("title/phrase").InnerText);

            Console.WriteLine(" [" + group.Attributes["size"].Value + " document(s)]");

            foreach (XmlNode subgroup in group.SelectNodes("group"))
            {
                PrintGroup(subgroup, level + 1);
            }
        }
    }
}

File: applications/carrot2-dcs/examples/java/src/main/java/org/carrot2/dcs/Examples.java

/*
 * Carrot2 project.
 *
 * Copyright (C) 2002-2016, Dawid Weiss, Stanisław Osiński.
 * All rights reserved.
 *
 * Refer to the full license file "carrot2.LICENSE"
 * in the root folder of the repository checkout or at:
 * http://www.carrot2.org/carrot2.LICENSE
 */

package org.carrot2.dcs;

import java.io.*;
import java.net.*;
import java.util.*;

import org.dom4j.*;
import org.dom4j.io.SAXReader;

/**
 * Contains several examples of clustering using the Document Clustering Server.
 */
final class Examples
{
    /** Shared path to the XML file to be clustered. */
    private static final String XML_FILE_PATH = "../shared/data-mining.xml";

    /** Shared XML feed. */
    private static final String XML_FEED = 
        "http://search.carrot2.org/stable/xml?source=web&type=CARROT2&q=test&results=20";

    /** Shared DCS address. */
    private static final URI dcsURI;
    static
    {
        try
        {
            dcsURI = new URI("http://localhost:8080/dcs/rest");
        }
        catch (URISyntaxException e)
        {
            throw new RuntimeException(e);
        }
    }

    /**
     * Anything capable of formatting, sending and parsing a multipart HTTP POST.
     */
    private final IHttpMultipartPostProvider httpPoster;

    /*
     * 
     */
    public Examples(IHttpMultipartPostProvider httpPoster)
    {
        this.httpPoster = httpPoster;
    }

    /**
     * Cluster data from an XML file (local).
     */
    public void clusterFromFile() throws IOException
    {
        final Map<String, String> attributes = new LinkedHashMap<String, String>();

        System.out.println("## Clustering documents from a local file");

        /*
         * Note the optional query attribute, we can provide it to avoid creation of
         * trivial clusters.
         */

        attributes.put("dcs.c2stream", new String(
            StreamUtils.readFullyAndClose(
                new FileInputStream(XML_FILE_PATH)), "UTF-8"));
        attributes.put("query", "data mining");

        displayResults(httpPoster.post(dcsURI, attributes));
    }

    /**
     * Cluster data from an external XML stream feed (providing an URL to that feed).
     */
    public void clusterFromRemoteXML() throws IOException
    {
        final Map<String, String> attributes = new LinkedHashMap<String, String>();

        System.out.println("## Clustering documents from a remote XML feed");
        attributes.put("dcs.source", "xml");
        attributes.put("dcs.algorithm", "stc");
        attributes.put("XmlDocumentSource.xml", XML_FEED);

        displayResults(httpPoster.post(dcsURI, attributes));
    }

    /**
     * Cluster data retrieved from a search engine or some other source registered in the
     * DCS as a document source.
     */
    public void clusterFromSearchEngine() throws IOException
    {
        final Map<String, String> attributes = new LinkedHashMap<String, String>();

        /*
         * For this request, we will pass some additional attributes to the default
         * algorithm and ask to skip the fetched documents in the output (retrieve
         * clusters only).
         */

        System.out.println("## Clustering search results from a search engine");

        // We use etools meta search engine input component.
        attributes.put("dcs.source", "etools");
        attributes.put("query", "test");
        attributes.put("results", "20");
        attributes.put("dcs.algorithm", "lingo");
        attributes.put("dcs.clusters.only", "true");

        // Some customized algorithm parameters.
        attributes.put("LingoClusteringAlgorithm.desiredClusterCountBase", "10");
        attributes.put("LingoClusteringAlgorithm.factorizationQuality", "LOW");
        attributes.put("LingoClusteringAlgorithm.factorizationFactory",
            "org.carrot2.matrix.factorization.PartialSingularValueDecompositionFactory");

        displayResults(httpPoster.post(dcsURI, attributes));
    }

    /**
     * Runs all examples.
     */
    public void runAllExamples() throws IOException
    {
        clusterFromFile();
        clusterFromRemoteXML();
        clusterFromSearchEngine();
    }

    /**
     * Run all examples with all HTTP POST providers.  
     */
    public static void main(String [] args) throws IOException
    {
        IHttpMultipartPostProvider [] providers = {
            new HttpClientPostProvider(),
            new JaxRsPostProvider()
        };

        for (IHttpMultipartPostProvider provider : providers)
        {
            new Examples(provider).runAllExamples();
        }
    }
    
    /**
     * Simple parsing and display of the response. This method uses dom4j for parsing XML,
     * feel free to use anything that comes handy.
     */
    @SuppressWarnings("unchecked")
    private static void displayResults(InputStream results) throws IOException
    {
        try
        {
            final SAXReader reader = new SAXReader();
            final Document document = reader.read(results);
            final Iterator<Element> i = document.getRootElement().elementIterator("group");
            while (i.hasNext())
            {
                final Element group = i.next();
                display(group, 1);
            }
            System.out.println();
        }
        catch (DocumentException e)
        {
            throw new IOException("Could not parse response: " + e.getMessage());
        }
        finally
        {
            if (results != null)
            {
                results.close();
            }
        }
    }

    /**
     * Display a single cluster and its sub-clusters.
     */
    @SuppressWarnings({"unchecked"})
    private static void display(Element group, int level)
    {
        final int recursiveDocumentCount = Integer.parseInt(group.attribute("size").getValue());
        final String label = group.element("title").elementText("phrase");

        for (int i = 0; i < level; i++) System.out.print("  ");
        System.out.println(label + " [" + recursiveDocumentCount + " document(s)]");
        final Iterator<Element> i = group.elementIterator("group");
        while (i.hasNext())
        {
            display(i.next(), level + 1);
        }
    }
}
